sim7000-async/src/modem/mod.rs

```diff
@@ -267,25 +270,26 @@ impl<'c, P: ModemPower> Modem<'c, P> {
         }
     }
 
-    async fn authenticate(&self, commands: &CommandRunnerGuard<'_>) -> Result<(), Error> {
-        let apn = match self.apn {
-            Some(apn) => apn.into(),
+    async fn authenticate(&mut self) -> Result<(), Error> {
+        let commands = self.commands.lock().await;
+        match &self.apn {
+            Some(_) => {}
             None => {
                 log::debug!("no default APN set, checking network for suggested APN.");
                 let (network_apn, _) = commands.run(cgnapn::GetNetworkApn).await?;
                 let Some(apn) = network_apn.apn else {
                     log::error!("no APN set");
                     return Err(Error::NoApn);
                 };
-                apn
+                self.apn = Some(apn.into());
             }
         };
 
-        log::info!("authenticating with apn {:?}", apn);
+        log::info!("authenticating with apn {:?}", self.apn);
 
         commands
             .run(cstt::StartTask {
-                apn,
+                apn: self.apn.clone().unwrap(),
```

Comment:
I realize that this unwrap is fine, but I still don't like it.

Try returning a reference to the APN from the match statement instead.
[Option::insert](https://doc.rust-lang.org/std/option/enum.Option.html#method.insert) could be useful here.